plugins/ksp/src/org/jetbrains/kotlin/ksp/symbol/impl/binary/KSAnnotationDescriptorImpl.kt

```diff
@@ -65,10 +73,36 @@ private fun <T> ConstantValue<T>.toValue(): Any? = when (this) {
     else -> value
 }
 
-fun AnnotationDescriptor.createKSValueArguments(): List<KSValueArgument> =
-    allValueArguments.map { (name, constantValue) ->
+fun AnnotationDescriptor.createKSValueArguments(): List<KSValueArgument> {
+    val presentValueArguments = allValueArguments.map { (name, constantValue) ->
         KSValueArgumentLiteImpl.getCached(
             KSNameImpl.getCached(name.asString()),
             constantValue.toValue()
         )
     }
+    val presentValueArgumentNames = presentValueArguments.map { it.name.asString() }
+    val argumentsFromDefault = (this.type.constructor.declarationDescriptor as? ClassDescriptor)?.constructors?.single()?.let {
+        it.getAbsentDefaultArguments(presentValueArgumentNames)
+    } ?: emptyList()
+    return presentValueArguments.plus(argumentsFromDefault)
+}
+
+fun ClassConstructorDescriptor.getAbsentDefaultArguments(excludeNames: Collection<String>): Collection<KSValueArgument> {
+    return this.valueParameters.filterNot { param -> excludeNames.contains(param.name.asString()) || !param.hasDefaultValue() }
+        .map { param ->
+            KSValueArgumentLiteImpl.getCached(
+                KSNameImpl.getCached(param.name.asString()),
+                param.getDefaultValue()
+            )
+        }
+}
+
+fun ValueParameterDescriptor.getDefaultValue(): Any? {
+    val psi = this.findPsi()
+    return when (psi) {
+        null -> TODO("resolve constants in class files")
```

Comment:
Can we make this just return null, rather than throwing an exception, for better usability?

Reply:
Sure I can do that, but I was wondering if null will be even more confusing. My original plan was to insert a special value to indicate this, but can't think of a good one.

Reply:
I'm worrying that throwing exception here would hinder the usability too much for projects using default annotation values from libraries, which is pretty common. Another solution might be making default value evaluated (and therefore exception thrown) *only* when reading that specific parameter.